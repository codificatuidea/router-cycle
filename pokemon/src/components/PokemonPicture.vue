<template>
  <div class="pokemon-container">
    <img class="hidden-pokemon" :src="imgSrc" />
    <img v-if="showPokemon" class="fade-in" :src="imgSrc" />
  </div>
</template>

<script>
export default {
  props: {
    pokemonId: {
      type: Number,
      required: true
    },
    showPokemon: {
      type: Boolean,
      required: true,
      default: false
    }
  },
  computed: {
    imgSrc() {
      return `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/${this.pokemonId}.png`
    }
  }
}
</script>

<style>
.pokemon-container {
    height: 200px;
}
img {
  height: 200px;
  position: absolute;
  margin-left: -70px;
  user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-user-drag: none;
  -webkit-user-select: none;
}
.hidden-pokemon {
    filter: brightness(0);
}
</style>