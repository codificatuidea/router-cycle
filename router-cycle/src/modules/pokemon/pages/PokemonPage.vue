<template>
  <h1>Pokemon #<span>{{ id }}</span></h1>
  <div v-if="pokemon">
    <img :src="pokemon.sprites.other.dream_world.front_default" />
    <h3>{{ pokemon.name }}</h3>
  </div>
  <!--<h1>Pokemon #<span>{{ this.$route.params.id }}</span></h1>-->
</template>

<script>
export default {
  props: {
    id: {
      required: true,
      type: Number
    }
  },
  data() {
    return {
      //id: null
      pokemon: null
    }
  },
  created() {
    //const { id } = this.$route.params
    //this.id = id
    this.getPokemon()
  },
  methods: {
    async getPokemon() {
      try {
        const request = await fetch(`https://pokeapi.co/api/v2/pokemon/${this.id}`)
        const response = await request.json()
        this.pokemon = response
      } catch (error) {
        this.$router.push('/')
      }
    }
  },
  watch: {
    id() {
      this.getPokemon()
    }
  }
}
</script>

<style>

</style>