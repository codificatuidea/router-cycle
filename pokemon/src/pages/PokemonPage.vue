<template>
  <h1 v-if="!pokemon">Espere por favor...</h1>
  <div v-else>
    <h1>¿Quién es ese Pokémon?</h1>
    <template v-if="showAnswer">
      <h2>{{ message }}</h2>
      <button @click="newGame">Nuevo Juego</button>
    </template>
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
    <PokemonOptions
      :pokemons="pokemons"
      @selection="checkAnswer"
    />

  </div>
</template>

<script>
import PokemonOptions from '@/components/PokemonOptions'
import PokemonPicture from '@/components/PokemonPicture'
import getPokemonOptions from '@/helpers/getPokemonOptions'

export default {
    components: {
        PokemonPicture,
        PokemonOptions
    },
    data() {
      return {
        pokemon: null,
        pokemons: [],
        showPokemon: false,
        message: '',
        showAnswer: false
      }
    },
    methods: {
      async getPokemons() {
        this.pokemons = await getPokemonOptions()
        const pokemonRandom = Math.floor(Math.random() * 4)
        this.pokemon = this.pokemons[pokemonRandom]
      },
      checkAnswer(pokemonFromSelection) {
        this.showPokemon = true
        this.showAnswer = true

        if (pokemonFromSelection.id === this.pokemon.id) {
          this.message = `Si, el pokémon es ${this.pokemon.name}`
        } else {
          this.message = `No, el pokémon es ${this.pokemon.name}`
        }
      },
      newGame() {
        this.showPokemon = false
        this.pokemons = []
        this.pokemon = null
        this.showAnswer = false
        this.getPokemons()
      }
    },
    mounted() {
      this.getPokemons()
    }
}
</script>

<style>

</style>