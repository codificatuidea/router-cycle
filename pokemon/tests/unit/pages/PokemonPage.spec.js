import { shallowMount } from "@vue/test-utils"
import PokemonPage from '@/pages/PokemonPage'
import { pokemons } from "../mocks/pokemons.mock"
import PokemonPicture from '@/components/PokemonPicture'

describe('PokemonPage component', () => {

    let wrapper
    let getPokemonsSpy

    beforeEach(() => {
        getPokemonsSpy = jest.spyOn(PokemonPage.methods, 'getPokemons')
        wrapper = shallowMount(PokemonPage)
    })

    test('Should do match with snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('Should call getPokemons on mount', () => {
        expect(getPokemonsSpy).toHaveBeenCalled()
    })

    test('Should do match with snapshot when pokemons are loaded', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                  pokemon: pokemons[0],
                  pokemons,
                  showPokemon: false,
                  message: '',
                  showAnswer: false
                }
            },
        })
        expect(wrapper.html()).toMatchSnapshot()
    })

    test.only('Should show PokemonPicture and PokemonOptions', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                  pokemon: pokemons[0],
                  pokemons,
                  showPokemon: false,
                  message: '',
                  showAnswer: false
                }
            },
        })
        //console.log(wrapper.find('pokemon-picture-stub'))
        //console.log(wrapper.findComponent(PokemonPicture))

        const pokemonPictureComponent = wrapper.findComponent(PokemonPicture)
        const pokemonPictureStub = wrapper.findComponent('pokemon-picture-stub')

        expect(pokemonPictureComponent.exists()).toBeTruthy()
        expect(pokemonPictureStub.exists()).toBeTruthy()

        expect(pokemonPictureComponent.props('pokemonId')).toBe(pokemons[0].id)
        expect(pokemonPictureStub.props('pokemonId')).toBe(pokemons[0].id)

    })

    test.only('Check answer', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                  pokemon: pokemons[0],
                  pokemons,
                  showPokemon: false,
                  message: '',
                  showAnswer: false
                }
            },
        })
        //console.log(wrapper.vm.checkAnswer(1))
    })

})