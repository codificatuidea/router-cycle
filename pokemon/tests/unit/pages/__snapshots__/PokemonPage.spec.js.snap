// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component Should do match with snapshot 1`] = `<h1>Espere por favor...</h1>`;

exports[`PokemonPage component Should do match with snapshot when pokemons are loaded 1`] = `
<div>
  <h1>¿Quién es ese Pokémon?</h1>
  <!--v-if-->
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
</div>
`;
